altapay.php

```diff
@@ -0,0 +1,2353 @@
+
+        foreach ($orderLines as $key => $orderedQuantity) {
+            if ($orderedQuantity > 0) {
+                $productDetails = $productDetail[$key];
+                if (!empty($productDetails)) {
+                    $productName = $productDetails['product_name'];
+
+                    $priceWithoutReductionTaxIncl = $productDetails['unit_price_tax_incl'] / (1 - ($productDetails['reduction_percent'] / 100));
+                    $basePrice = $productDetails['original_product_price'];
+
+
+                    $cartRuleDiscounts = $this->getCartRuleDiscounts($orderDetail);
+                    //Calculation of base price
+
+                    if ($productDetails['reduction_percent'] > 0) {
+                        $discountPercentage = $productDetails['reduction_percent'];
+
+                    } else if (!empty($cartRuleDiscounts)) {
+                        foreach ($cartRuleDiscounts as $cartRuleDiscount) {
+                            if ($productDetails['product_id'] == $cartRuleDiscount['productID']) {
+                                $discountPercentage = $cartRuleDiscount['discountPercent'];
+                                break;
+                            } else {
+                                $discountPercentage = 0;
+                            }
+
+                        }
+                    }
+
+                    if ($productDetails['product_attribute_id']) {
+                        $itemID = $productDetails['product_reference'] . '-' . $productDetails['product_attribute_id'];
+                    } else {
+                        $itemID = $productDetails['product_reference'];
+                    }
+
+                    $productQuantity = $orderedQuantity;
+                    $productTax = $priceWithoutReductionTaxIncl - $basePrice;
+                    $goodsType = 'item';
+
+                    if ($goodWillRefund) {
+                        $goodsType = 'refund';
+                    }
+
+                    //Looping into the product array to get the difference regarding compensation amount
+                    foreach ($productDetail as $proKeys) {
+                        $productPriceTaxIncl = $proKeys['total_price_tax_incl'];
+                        $priceAfterDiscountRounded += (float)number_format($productPriceTaxIncl - ($productPriceTaxIncl * ($discountPercentage / 100)), 2, '.', '');
+                        $priceAfterDiscount += $productPriceTaxIncl - ($productPriceTaxIncl * ($discountPercentage / 100));
+                        $totalQuantity += $proKeys['product_quantity'];
+                    }
+                    //Calculation of Total Compensation Amount
+                    $compensationAmount = (float)number_format($priceAfterDiscountRounded - $priceAfterDiscount, 2, '.', '');
+                    $compensationAmountPerQuantity = $compensationAmount / $totalQuantity;
+                    $totalProductsTaxAmount = (float)number_format($productTax * $productQuantity, 2, '.', '');
```

Comment:
maybe use `round()` instead of this

Reply:
I have already tried it with round(), round basically gives you either floor or ceiling values. Whereas we need the real value with number_format at specific decimal. That is why we are using number_format

Reply:
The second parameter of `round` allows you to specify what precision you want.
```suggestion
                    $totalProductsTaxAmount = round($productTax * $productQuantity, 2);
```

Reply:
Yes but still considering the precision to round either ceil or floor is the reason we are using number_format. Beside this there is a scenario where in different countries, the decimal and separator conditions are different.

Reply:
`round` and `number_format` perform the same logic regarding precision. Since you are performing a multiplication before calling `number_format` and are casting to float right after you are not handling varying decimal separators. In fact the current code would fail in that case as you are forcing it to a period and then trying to cast the resulting string to a native float.

Reply:
When they are performing with same logic and number_format is giving us the ability to set our defined separator, then I think its fine to keep it to number_format. It is not failing in any case as for now but for the case of casting it float, we can remove the explicit casting.

Reply:
Removing the cast to `float` makes sense for lines `862` and `870` since they are sent in the API request, doing so would indeed make sure they are formatted correctly regardless of the system locale.

But line `860` and `855` should be changed to use `round()` as their values are used with math operators.